document/AnnexE-Determinism

```diff
@@ -0,0 +1,199 @@
+For example, within the `file.hashes` property, SARIF does not require the `hash` objects
+to appear in any particular order.
+For such arrays, a tool can ensure the order by sorting the array elements before
+writing them to the log file.
+For example, it might sort the `hash` objects alphabetically by the string value of the
+`hash.algorithm` property.
+
+A tool might similarly choose to emit the string elements of a `properties.tags` array in
+locale-insensitive alphabetical order.
+
+The array of `result` objects presents more of a problem.
+A multi-threaded analysis tool analyzing multiple files in parallel might produce results
+in any order, and there is no natural order for the results.
+A tool might choose to order them, for example, first alphabetically by analysis target URI,
+then numerically by line number, then by column number, then alphabetically by rule id.
+
+For dictionaries such as the `run.rules` object or the `run.files` object,
+a tool might order the property names alphabetically, using a locale-insensitive ordering.
+
+## Absolute paths { - #Annex-E-AbsolutePaths }
+
+The use of absolute file paths in URI-valued properties such as `physicalLocation.uri`
+makes it difficult to produce deterministic output.
+For example:
+
+* Different build machines might be configured to use different source directories.
+* A single build machine might use a different directory for each build.
+
+For a tool to produce deterministic output, it must avoid the use of absolute file paths.
+Tools can achieve this by emitting URIs that are relative to one or more root directories
+(for example, a source root directory and an output root directory),
+and accompanying each URI-valued property with a URI base id property (&sect;[#fileFormat-uriBaseIdProperties]). 
+
+## Case-insensitive file format elements { - #Annex-E-Case }
+
+Case-insensitive file systems can also cause non-deterministic output.
+Suppose a URI-valued property contains a file path from a file system that is
+case-preserving but case-insensitive (such as NTFS):
+
+```
+"uri": "file:///C:/Code/File1.c"
+```
+
+Suppose the file is then renamed to `file.c`. NTFS treats `File1.c` and `file1.c` as the same file.
+But a subsequent run of the tool might produce
+
+```
+"uri": "file:///C:/Code/file1.c"
+```
+
+A tool might choose to ignore this issue, since this would happen rarely, and the only
+consequence of ignoring it is that a subsequent step in a build process might execute
+even though it does not strictly have to.
+
+To produce absolutely deterministic output, the tool might instead choose to convert all
+file paths to lowercase or to uppercase, using a locale-insensitive case conversion:
+
+```
+"uri": "file:///C:/code/file1.c"
+```
+
+~IsoNote
+
+On the .NET platform, the recommendation is to convert strings to uppercase
+using the method `String.ToUpperInvariant`, which would result in:
+
+```
+"uri": "file:///C:/CODE/FILE1.C"
+```
+~
```

Comment:
ok, let's open an issue. as per other comments, i'd prefer we put language in on the security concerns that drive this practice to allow people to make a good decision. i.e., we don't want to advise upper-casing as a general practice, i think. it would be unusual to require it.

---
In reply to: [67031992](https://github.com/sarif-standard/sarif-spec/pull/223#discussion_r67031992) [](ancestors = 67031992)